designs/multiple-worker-node-groups.md

```diff
@@ -0,0 +1,241 @@
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: eksa-test-worker-node-template-1638469395669
+  namespace: eksa-system
+spec:
+  template:
+    spec:
+    ...
+---
+apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+kind: KubeadmConfigTemplate
+metadata:
+  name: eksa-test-2-md-0
+  namespace: eksa-system
+spec:
+  template:
+    spec:
+      joinConfiguration:
+        pause:
+          imageRepository: public.ecr.aws/eks-distro/kubernetes/pause
+          imageTag: v1.20.7-eks-1-20-8
+        bottlerocketBootstrap:
+          imageRepository: public.ecr.aws/l0g8r8j6/bottlerocket-bootstrap
+          imageTag: v1-20-8-eks-a-v0.0.0-dev-build.579
+        nodeRegistration:
+          criSocket: /var/run/containerd/containerd.sock
+          kubeletExtraArgs:
+            cloud-provider: external
+            tls-cipher-suites: Something
+          name: '{{ ds.meta_data.hostname }}'
+	  ...
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachineDeployment
+metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: eksa-test
+  name: eksa-test-2-md-0
+  namespace: eksa-system
+spec:
+  clusterName: eksa-test
+  replicas: 3
+  selector:
+    matchLabels: {}
+  template:
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: eksa-test
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfigTemplate
+          name: eksa-test-2-md-0
+      clusterName: eksa-test
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: VSphereMachineTemplate
+        name: test
+      version: v1.20.7-eks-1-20-8
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: test
+  namespace: eksa-system
+spec:
+  template:
+    spec:
+    ...
+---
+```
+
+For each worker node groups, CAPI spec file will continue to have the following 3 kind fields.
+
+* KubeadmConfigTemplate
+* Machime deployment
+* VSphereMachineTemplate
+
+For each group, we will append these three fields corresponding to that group in the CAPI spec.
+
+Right now, the cli assumes that there will be only one group and it treats worker node group configuration array as a collection of only one element. As a result, the controller just refers to the first element of this array in different places of the code. So we need to do the same operations in loops, which includes CAPI spec creation, cluster spec validation etc. Once a CAPI spec is created with this approach, the workload cluster will be created with multiple worker nodes.
+
+Also, it needs to be made sure that at the least one of the worker node groups does not have `NoExecute` or `NoSchedule` taint. This validation will be done at the preflight validation stage.
+
+The examples in this design are for vsphere provider. But the same strategy applies for other providers as well.
```

Comment:
This means that for docker, only changing the taints for each worker node group would warrant a new capi template spec, or just each worker node group corresponds to a separate capi template spec even if the values are exactly the same?

Reply:
Yes, for docker also, we will be adding KubeadmConfigTemplate, MachineDeployment, VSphereMachineTemplate in the CAPI spec file for each worker node group.